### DISCUSSION TITLE
when i updated the dag, task in task_instance didn't update

### DISCUSSION THREAD

**Discussion by broholens** [apache/airflow#24319] [General]
### Apache Airflow version

2.3.2 (latest released)

### What happened

There is a dag with pool1，and task in task_instance of this dag works on pool1，when i update the pool of the dag to pool2,  task in task_instance of this dag still works on pool1.

I checked `serialized_dag` table of this dag, and the data already updated to pool2. 

So i read the code, and i find there may have a bug in function `_add_dag_from_db` of `dagbay.py`.  And it happens in a very low probability.
```python
    def _add_dag_from_db(self, dag_id: str, session: Session):
        """Add DAG to DagBag from DB"""
        from airflow.models.serialized_dag import SerializedDagModel

        row = SerializedDagModel.get(dag_id, session)
        if not row:
            return None

        row.load_op_links = self.load_op_links
        dag = row.dag
        for subdag in dag.subdags:
            self.dags[subdag.dag_id] = subdag
        self.dags[dag.dag_id] = dag
        self.dags_last_fetched[dag.dag_id] = timezone.utcnow()
        self.dags_hash[dag.dag_id] = row.dag_hash
```
Let's say there's a situation：
when `row = SerializedDagModel.get(dag_id, session)` executed and `self.dags_last_fetched[dag.dag_id] = timezone.utcnow()`  not executed, other process update the table `serialized_dag` of this dag,  `self.dags` will store the old dag, and self.dags_last_fetched will be bigger than `serialized_dag`  update_time，then code below will never execute:
```python
if sd_last_updated_datetime > self.dags_last_fetched[dag_id]:
    self._add_dag_from_db(dag_id=dag_id, session=session)
```
#### so the dag in memory will never be updated.

And i think this code will fix the bug:
```python
    def _add_dag_from_db(self, dag_id: str, session: Session):
        """Add DAG to DagBag from DB"""
        from airflow.models.serialized_dag import SerializedDagModel
        
        last_fetched_time = timezone.utcnow()
        row = SerializedDagModel.get(dag_id, session)
        if not row:
            return None

        row.load_op_links = self.load_op_links
        dag = row.dag
        for subdag in dag.subdags:
            self.dags[subdag.dag_id] = subdag
        self.dags[dag.dag_id] = dag
        self.dags_last_fetched[dag.dag_id] = last_fetched_time
        self.dags_hash[dag.dag_id] = row.dag_hash
```

### Operating System

linux

### Deployment

Official Apache Airflow Helm Chart

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by potiuk** [comment]
Can you please make a PR with a fix and unit test reproducing the case? I think this will be much easier to discuss about it when we see the PR and can discuss on actual code. The worst case - we will close the PR.

**Comment by potiuk** [comment]
Converted it into a discussion as well, until we know if this is really a bug.
